File: test_totp.js
const crypto = require('crypto');

function generateTOTP(secret, window = 0) {
    const base32 = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567';
    let bits = '';
    for (let i = 0; i < secret.length; i++) {
        const val = base32.indexOf(secret.charAt(i).toUpperCase());
        bits += val.toString(2).padStart(5, '0');
    }
    const key = Buffer.alloc(Math.floor(bits.length / 8));
    for (let i = 0; i < key.length; i++) {
        key[i] = parseInt(bits.substring(i * 8, i * 8 + 8), 2);
    }
    
    const epoch = Math.round(Date.now() / 1000.0);
    const time = Buffer.alloc(8);
    let counter = Math.floor(epoch / 30) + window;
    time.writeUInt32BE(Math.floor(counter / 0x100000000), 0);
    time.writeUInt32BE(counter & 0xffffffff, 4);

    const hmac = crypto.createHmac('sha1', key);
    hmac.update(time);
    const digest = hmac.digest();
    
    const offset = digest[digest.length - 1] & 0xf;
    const code = (digest.readUInt32BE(offset) & 0x7fffffff) % 1000000;
    return code.toString().padStart(6, '0');
}

const secret = "KG327CRMRTJSQDLFYUTE4MVI4LQZUBCF";
console.log("Current TOTP code:", generateTOTP(secret));

async function testSolcast() {
    const userName = "adsi.om@alterpowerdigos.com";
    const Database = require('better-sqlite3');
    const db = new Database("C:\\ProgramData\\InverterDashboard\\db\\adsi.db", { readonly: true });
    const row = db.prepare("SELECT value FROM settings WHERE key='solcastToolkitPassword'").get();
    const password = row.value;
    db.close();

    const authUrl = "https://api.solcast.com.au/auth/credentials";
    const totpCode = generateTOTP(secret);
    console.log("Submitting login with TOTP:", totpCode);
    
    const resp = await fetch(authUrl, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
            "Accept": "application/json"
        },
        body: JSON.stringify({
            userName,
            password,
            rememberMe: true,
            meta: { TwoFactorCode: totpCode }
        })
    });
    
    console.log("HTTP Status:", resp.status);
    console.log("Response Text:", await resp.text());
    console.log("Cookies:", resp.headers.get("set-cookie"));
}

testSolcast().catch(console.error);

File: test_totp.py
import sqlite3
import hmac
import hashlib
import time
import base64
import struct
import json
import requests

def get_totp_token(secret: str) -> str:
    # Decode base32 secret
    # Pad to multiple of 8
    padding = '=' * ((8 - len(secret) % 8) % 8)
    key = base64.b32decode(secret + padding, casefold=True)
    
    # 30 second window
    timestamp = int(time.time()) // 30
    msg = struct.pack(">Q", timestamp)
    
    # HMAC-SHA1
    h = hmac.new(key, msg, hashlib.sha1).digest()
    o = h[19] & 15
    token = (struct.unpack(">I", h[o:o+4])[0] & 0x7fffffff) % 1000000
    return f"{token:06d}"

def test_login():
    db_path = r"C:\ProgramData\InverterDashboard\db\adsi.db"
    conn = sqlite3.connect(f"file:{db_path}?mode=ro", uri=True)
    c = conn.cursor()
    c.execute("SELECT value FROM settings WHERE key='solcastToolkitPassword'")
    row = c.fetchone()
    if not row:
        print("Password not found")
        return
    password = row[0]
    
    email = "adsi.om@alterpowerdigos.com"
    secret = "KG327CRMRTJSQDLFYUTE4MVI4LQZUBCF"
    totp = get_totp_token(secret)
    print(f"Generated TOTP: {totp}")
    
    url = "https://api.solcast.com.au/auth/credentials"
    
    # Test JSON submission
    payload = {
        "userName": email,
        "password": password,
        "rememberMe": "true",
        "meta": { "TwoFactorCode": totp }
    }
    
    headers = {
        "Content-Type": "application/json",
        "Accept": "application/json"
    }
    
    print("Sending POST...")
    resp = requests.post(url, json=payload, headers=headers)
    print("Status:", resp.status_code)
    print("Response:", resp.text[:200])
    print("Cookies:", resp.cookies.get_dict())

if __name__ == "__main__":
    test_login()
